controller/api/destination/watcher/ip_watcher.go

```diff
@@ -0,0 +1,372 @@
+func (iw *IPWatcher) Unsubscribe(clusterIP string, port Port, listener EndpointUpdateListener) {
+	iw.log.Infof("Stopping watch on service cluster ip [%s:%d]", clusterIP, port)
+	ss, ok := iw.getServiceSubscriptions(clusterIP)
+	if !ok {
+		iw.log.Errorf("Cannot unsubscribe from unknown service ip [%s:%d]", clusterIP, port)
+		return
+	}
+	ss.unsubscribe(port, listener)
+}
+
+func (iw *IPWatcher) addService(obj interface{}) {
+	service := obj.(*corev1.Service)
+	if service.Namespace == kubeSystem || service.Spec.ClusterIP == "None" {
+		return
+	}
+
+	ss := iw.getOrNewServiceSubscriptions(service.Spec.ClusterIP)
+
+	ss.updateService(service)
+}
+
+func (iw *IPWatcher) deleteService(obj interface{}) {
+	service := obj.(*corev1.Service)
+	if service.Namespace == kubeSystem {
+		return
+	}
+
+	ss, ok := iw.getServiceSubscriptions(service.Spec.ClusterIP)
+	if ok {
+		ss.deleteService()
+	}
+}
+
+func (iw *IPWatcher) addPod(obj interface{}) {
+	pod := obj.(*corev1.Pod)
+	if pod.Namespace == kubeSystem {
+		return
+	}
+	if pod.Status.PodIP == "" {
+		// Pod has not yet been assigned an IP address.
+		return
+	}
+	ss := iw.getOrNewServiceSubscriptions(pod.Status.PodIP)
+
+	ownerKind, ownerName := iw.k8sAPI.GetOwnerKindAndName(pod, true)
+	podSet := PodSet{
+		Pods: map[PodID]Address{
+			PodID{
+				Name:      pod.Name,
+				Namespace: pod.Namespace,
+			}: Address{
+				IP:        pod.Status.PodIP,
+				Port:      0, // Will be set by individual subscriptions
+				Pod:       pod,
+				OwnerName: ownerName,
+				OwnerKind: ownerKind,
+			},
+		},
+		Labels: map[string]string{},
+	}
+
+	ss.updatePod(podSet)
+}
+
+func (iw *IPWatcher) deletePod(obj interface{}) {
+	pod := obj.(*corev1.Pod)
+	if pod.Namespace == kubeSystem {
+		return
+	}
+
+	ss, ok := iw.getServiceSubscriptions(pod.Status.PodIP)
+	if ok {
+		ss.deletePod()
+	}
+}
+
+// Returns the serviceSubscriptions for the given clusterIP if it exists.  Otherwise,
+// create a new one and return it.
+func (iw *IPWatcher) getOrNewServiceSubscriptions(clusterIP string) *serviceSubscriptions {
+	iw.Lock()
+	defer iw.Unlock()
+
+	// If the service doesn't yet exist, create a stub for it so the listener can
+	// be registered.
+	ss, ok := iw.publishers[clusterIP]
+	if !ok {
+		ss = &serviceSubscriptions{
+			clusterIP: clusterIP,
+			listeners: make(map[EndpointUpdateListener]Port),
+			endpoints: iw.endpoints,
+			log:       iw.log.WithField("clusterIP", clusterIP),
+		}
+
+		objs, err := iw.k8sAPI.Svc().Informer().GetIndexer().ByIndex(podIPIndex, clusterIP)
+		if err != nil {
+			iw.log.Error(err)
+		} else {
+			if len(objs) > 1 {
+				iw.log.Errorf("Service cluster IP conflict: %v, %v", objs[0], objs[1])
+			}
+			if len(objs) == 1 {
+				if svc, ok := objs[0].(*corev1.Service); ok {
+					ss.service = ServiceID{
+						Namespace: svc.Namespace,
+						Name:      svc.Name,
+					}
+				}
+			}
+		}
+		objs, err = iw.k8sAPI.Pod().Informer().GetIndexer().ByIndex(podIPIndex, clusterIP)
```

Comment:
the endpoints watcher has already added the pod ip indexer so we don't need to do it.

I'm using the name `clusterIP` a bit more broadly to mean the ip address of a resource in the cluster.  This can be the ip address of a pod, or the cluster ip of a service.